### DISCUSSION TITLE
Custom controller page performance

### DISCUSSION THREAD

**Discussion by scottmetoyer** [statamic/cms#6913] [Q&A] [answered]
I have a few routes set up to use custom controllers in my Statamic 3.3 site hosted on Forge.

I've noticed that performance for the pages that are using custom controllers is very slow, almost like those pages are not using any caching. I have a very simple controller set up that just grabs and entry and returns a view that dumps the entry to the screen. This page takes up to 10 seconds to load while the same entry when viewed through the default Statamic routes loads in just a few hundred ms.

Any ideas what may be going on?

**Comment by jasonvarga** [comment]
Can you show the code?

**Comment by scottmetoyer** [comment]
Here's one controller that is showing the performance issue:

```
public function version($slug)
{
   $entry = Entry::findBySlug($slug, 'products');
   
   return (new \Statamic\View\View)
      ->template("products.version-history")
      ->layout('layout')
      ->cascadeContent($entry)
       ->with([
           'title' => $entry->title . " Version History",
        ]);
}
```

Here's the template:

```
{{ partial src="bit/product/nav" }}

<section class="flex items-center justify-center py-8 bg-center bg-cover" style="background-image: url({{ backgrounds.0.url }});">
    <div class="container px-4 mx-auto space-y-4 text-center md:px-0">
        <h1 class="text-3xl font-bold md:text-5xl {{ backgrounds.0.image_text_color ?? text_color }}">{{ title }}</h1>
    </div>
</section>

<section id="info" class="py-12 bg-gray-50 dark:bg-gray-750">
    <div class="container px-4 mx-auto space-y-8 md:px-0">
        <div class="mx-auto space-y-4 max-w-prose">
            {{ partial src="bit/content" :content="version_history" }}
        </div>
    </div>
</section>
```

The content partial looks like this:

```
{{ content }}
    {{ partial src="sets/{type}" :light="light" :size="size" :width="width" }}
{{ /content }}
```

**Comment by scottmetoyer** [comment]
I should also mention I'm using the 'half' caching strategy. I double checked that the URLs to the slow pages are not in the exclude array.

I turned on debugbar and noticed that the pages with custom controllers don't appear to be cached - debugbar shows views and messages only for those slower pages. 

If I turn off half caching all of the pages perform similarly slow and debugbar shows messages and views for all pages on the site. Is there a trick to getting Statamic static caching working with custom controllers?

**Reply by jackmcdade** [reply]
Can you test without Debugbar? It enables a bunch of stuff and creates more work in the background – i wonder if that's actually part of the problem?

**Reply by scottmetoyer** [reply]
I have the issue with or without Debugbar activated. I only turned it on to see if I could get a clue to what was going on. Thanks for looking and the idea though!

**Comment by jasonvarga** [comment]
> the same entry when viewed through the default Statamic routes loads in just a few hundred ms.

You've mentioned you're using static caching.

If you disable static caching and load one of the regular Statamic routes, is it just as slow as your custom route?

**Comment by scottmetoyer** [comment]
If I disable static caching, all pages load equally slow. If I enable it, only the custom controller routes load slow.

It definitely appears that static caching doesn’t work for custom controller routes, but I’m not sure if that’s by design.

We recently upgraded the Forge server to PHP 8.1 from 7.4 which may also be a clue.

Scott Metoyer
***@***.***

**Comment by jasonvarga** [comment, marked as answer]
Clearly there's an issue regarding the load times, but in order to enable static caching on your custom route, you should get it to use the cache middleware.

```php
Route::get('...')->middleware([
  Statamic\StaticCaching\Middleware\Cache::class
]);
```

**Reply by scottmetoyer** [reply]
Nice, I will try that - thanks!

Regarding load times, I noticed many copies of this warning message in Debugbar:

`[03:02:48] LOG.warning: strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated in /vendor/laravel/framework/src/Illuminate/Support/Str.php on line 839`

It's repeated over 900 times and seems to have sprung up after upgrading to PHP 8.1. I'm not sure if it's related.
